### DISCUSSION TITLE
Indicator that someone is watching

### DISCUSSION THREAD

**Discussion by danielmarschall** [rustdesk/rustdesk#6451] [Feature Request]
It would be great if there could be an indicator that someone is currently watching.

In TeamViewer and anydesk, the missing wallpaper was my indicator (#1586), however, this has the disadvantage that I don't notice it when I am working on both screens with full-screen applications.

It would be better to have a small "flag" at the side of the screen, like TeamViewer has, or a rectangle (like a camera HUD) around the screen like FastViewer.

It would be amazing if that indicator could also show me if the partner is watching screen A or screen B. (Often, you explain something over the phone and the person watches on the wrong screen and wonders what I am talking about)

Note that if multiple people are watching, it might be good if there would be multiple flags on the screen(s), so you could see who is watching on which screen.

Would this be possible?

**Comment by danielmarschall** [comment]
To introduce the feature for myself,
I have created a small Delphi program that runs in the background and searches for a connection incoming window (which stays during the session).
It then shows a small blinking window that is set as "TopMost". This way I can always see that someone is currently watching on my Desktop.

![image](https://github.com/rustdesk/rustdesk/assets/28412477/1dab51d2-1f6c-48d9-94a3-4958472de19f)

This small window automatically closes on the connection window is closed (which happens when the session is closed).

```
uses
  DateUtils;

function EnumWindowsProc(wHandle: HWND; Form: TForm1): BOOL; stdcall;
var
  Title, ClassName: array[0..255] of char;
begin
  GetWindowText(wHandle, Title, 255);
  GetClassName(wHandle, ClassName, 255);
  if (Pos(' - RustDesk', Title) >= 1) and  // Incoming connection: In front of the hyphen there is the ID number of the remote side
     (Pos('- Remote Desktop - RustDesk', Title) = 0) then  // Outgoing connection: In front of the hyphen there is the ID number of the remote side
  begin
    Form.Tag := 2;

    // FIXES https://github.com/rustdesk/rustdesk/discussions/6452
    //       "Connection accept window should be at the front"
    SetWindowPos(wHandle, HWND_TOPMOST, 0, 0, 0, 0, SWP_NOMOVE or SWP_NOSIZE or SWP_SHOWWINDOW);
  end;
  Result := True;
end;

procedure TForm1.BitBtn1Click(Sender: TObject);
begin
  Visible := false;
end;

procedure TForm1.FormCreate(Sender: TObject);
begin
  Top := 0;
  Width := Screen.Width div 2;
  Left := (Screen.Width div 2) - (Width div 2);
  Height := 25;
  BitBtn1.Width := 25;
end;

procedure TForm1.FormShow(Sender: TObject);
begin
  // Hide form from Taskbar
  ShowWindow(Application.Handle,SW_HIDE); // Important: Application.MainFormOnTaskbar := False!
end;

procedure TForm1.Timer1Timer(Sender: TObject);
resourcestring
  LNG_INCOMING_CONN = 'Aktive eingehende RustDesk-Verbindung !';
  LNG_INCOMING_CONN_DONE = 'Aktive eingehende RustDesk-Verbindung !';
begin
  if Tag = 0 then Visible := false; // Start with invisible

  Tag := 1;
  EnumWindows(@EnumWindowsProc, LPARAM(Self));
  BitBtn1.Visible := Tag = 1;
  if Tag = 2 then
  begin
    // FIXES https://github.com/rustdesk/rustdesk/discussions/6451
    //       "Indicator that someone is watching"

    Visible := true; // stay visible to see if somebody WAS connected while you were away!
    Panel1.Caption := LNG_INCOMING_CONN;
    if DateTimeToUnix(Now) mod 2 = 0 then
      Panel1.Color := clYellow
    else
      Panel1.Color := clRed;
  end
  else
  begin
    Panel1.Caption := LNG_INCOMING_CONN_DONE;
    Panel1.Color := clLime;
  end;
end;
```

**Reply by jtagcat** [reply]
Leaving it as an option defeats the purpose, but I do find the yellow bar for example distracting and possibly obstructing. With AnyDesk, users have gotten used to the 'if red icon (often with orange background) present on toolbar = watching; no window = no connection).

**Comment by rustdesk** [comment]
I totally agree with you, this is a good feature, but it won't be in our high priority, no ETA yet.

**Reply by danielmarschall** [reply]
Thank you :-) I am glad you like this proposal!
I have no problem with waiting for that feature.
I am super happy that RustDesk exists and you do a great job making it so stable and reliable.

**Comment by pixelstuff** [comment]
On Windows could this indicator just be shown as a separate connection icon on the taskbar with a unique color that would make it obvious a connection was active? If more than one person connected then more than one icon is displayed. I don't know if the Windows API allows for that. Then of course one could hover over each of the icons to get more details about who is connected.
